chain-tx-enclave-next/mls/src/ciphersuite.rs

```diff
@@ -296,15 +292,88 @@ impl CipherSuite {
     }
 
     /// TODO: use generic array?
-    pub fn hash(&self, data: &[u8]) -> Vec<u8> {
+    pub fn hash(self, data: &[u8]) -> Vec<u8> {
         match self {
             CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => Sha256::digest(data).to_vec(),
         }
     }
 
-    pub fn hash_len(&self) -> usize {
+    pub fn hash_len(self) -> usize {
         match self {
             CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => 32,
         }
     }
+
+    /// spec: draft-ietf-mls-protocol.md#key-schedule
+    pub fn expand_label(
+        self,
+        secret: &[u8],
+        group_context_hash: Vec<u8>,
+        label: &str,
+        context: &[u8],
+        length: u16,
+    ) -> Result<Vec<u8>, InvalidLength> {
+        match self {
+            CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => Hkdf::<Sha256>::new(
+                None, secret,
+            )
+            .expand_label(group_context_hash, label, context, length),
+        }
+    }
+
+    pub fn secret_size(self) -> u16 {
```

Comment:
key_pair secret?

Reply:
The [go-mls](https://github.com/cisco/go-mls/blob/master/crypto.go#L69) used the name `SecretSize`, so

Reply:
so what's this `secret_size` if not key pair secret size? the output hash length?

Reply:
It’s used by expand label “path”, when generating path secrets.

Reply:
https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#ratchet-tree-evolution

`KEM.Nsk` -- Nsk is the length of the secret key, so should be the same `keypair_secret_size`?

Reply:
some of this changed, it may be good to put Git commit sha of spec this refers to

Reply:
"Each ciphersuite has a `Derive-Key-Pair` function that maps octet strings of
length `Nsk` (a KEM-specific constant defined by HPKE) to HPKE key pairs", it seems they are still different?
And in go-mls, the [P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L108) has SecretSize 64, but the [DHKEM_P256](https://github.com/cisco/go-mls/blob/master/crypto.go#L292) has keyPairSecretSize 32.

Reply:
> some of this changed, it may be good to put Git commit sha of spec this refers to

update the sha in lib.rs to refer to current master?

Reply:
> "Each ciphersuite has a `Derive-Key-Pair` function that maps octet strings of
> length `Nsk` (a KEM-specific constant defined by HPKE) to HPKE key pairs", it seems they are still different?
> And in go-mls, the [P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L108) has SecretSize 64, but the [HPKEKEM](https://github.com/cisco/go-mls/blob/master/crypto.go#L292) has keyPairSecretSize 32.

66, not 32: https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#key-encapsulation-mechanisms-kems-kem-ids

`shared_secret_size`? I'd need to read through more in detail whether that context is correct

Reply:
as HPKE library doesn't implement P-521 or X448 yet, I'm not sure if it even distinguishes these sizes :) but still good to keep it consistent and distinguish them if that's in the spec

Reply:
oh, I guess they have a bug here to [use DHKEM_P256 in ciphersuite P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L109)